src/lib.rs

```diff
@@ -0,0 +1,559 @@
+                node = self.nodes[index].parent;
+            }
+        }
+    }
+
+    fn get_node(&self, value: &T) -> Result<usize, Option<usize>> {
+        let mut current_index = self.root;
+        let mut previous_index = None;
+        if !self.nodes.is_empty() {
+            while let Some(index) = current_index {
+                previous_index = current_index;
+                if *value > self.nodes[index].value {
+                    current_index = self.nodes[index].right_child;
+                } else if *value < self.nodes[index].value {
+                    current_index = self.nodes[index].left_child;
+                } else {
+                    return Ok(index);
+                }
+            }
+        }
+        Err(previous_index)
+    }
+
+    fn create_node(&mut self, value: T) -> Result<Node<T>, usize> {
+        match self.get_node(&value) {
+            Ok(index) => {
+                Err(index)
+            },
+            Err(previous_index) => {
+                if let Some(parent) = previous_index {
+                    if self.nodes[parent].value < value {
+                        self.nodes[parent].right_child = Some(self.nodes.len());
+                    } else {
+                        self.nodes[parent].left_child = Some(self.nodes.len());
+                    }
+                }
+                Ok(Node {
+                    value: value,
+                    parent: previous_index,
+                    left_child: None,
+                    right_child: None,
+                    height: 0
+                })
+            }
+        }
+    }
+
+    fn update_heights(&mut self, start_index: usize) {
+        if !self.nodes.is_empty() {
+            let mut child_index = start_index;
+            while let Some(ancestor_index) = self.nodes[child_index].parent {
+                self.nodes[ancestor_index].height =
+                    self.compute_height(self.nodes[ancestor_index].left_child,
+                                        self.nodes[ancestor_index].right_child);
+                child_index = ancestor_index;
+            }
+            // update the height of the root
+            self.nodes[self.root.unwrap()].height =
+                self.compute_height(self.nodes[self.root.unwrap()].left_child,
+                                    self.nodes[self.root.unwrap()].right_child);
+        }
+    }
+
+    /// Helper function that computes the height of a node based
+    /// on the heights of its children.
+    fn compute_height(&self, l_subtree: Option<usize>,
+            r_subtree: Option<usize>) -> usize {
+        let (mut left_height, mut right_height) = (0, 0);
+        if let Some(index) = l_subtree {
+            left_height = self.nodes[index].height + 1;
+        }
+        if let Some(index) = r_subtree {
+            right_height = self.nodes[index].height + 1;
+        }
+        std::cmp::max(left_height, right_height)
+    }
+
+    fn balance_factor(&self, node: &Node<T>) -> i8 {
```

Comment:
Why does this return a signed int?

Reply:
The balance factor can be negative. In my code the balance factor is defined as `height(left_child) - height(right_child)`. If it is negative and less than -2, the tree is 'right-heavy'. If it's positive and greater than 2, the tree is 'right-heavy'

Reply:
Aha, got it. But can the balance exceed +/- 128?

Reply:
It should always be in the range `[-2, 2]`. If at some point it becomes `+/- 2`, the tree is (*should be*) rebalanced

Reply:
Understood... but how far can it exceed +/- 2 before it's rebalanced?

Reply:
>  In an AVL tree, the heights of the two child subtrees of any node differ by at most one
(wikipedia)

The 2 and -2 cases trigger rebalancing. You can't have `< -2` or `> 2`  because you only ever insert/delete one node at a time.

Reply:
Aha, got it.